File: src/exec.ts
import { spawn } from "node:child_process"
import { CommandExecutionError } from "./errors.js"
import type { CommandResult, RuntimeConfig } from "./types.js"

const SECRET_PATTERNS = [
  /AKIA[0-9A-Z]{16}/g,
  /ASIA[0-9A-Z]{16}/g,
  /(?<=aws_secret_access_key\s*=\s*)[^\s]+/gi,
  /(?<=aws_session_token\s*=\s*)[^\s]+/gi,
  /(?<=Authorization:\s*Bearer\s+)[A-Za-z0-9._~+/=-]+/gi,
  /(?<=token: )[A-Za-z0-9._~+/=-]+/gi
] as const

type ExecuteInput = {
  readonly command: string
  readonly args: readonly string[]
  readonly config: RuntimeConfig
  readonly extraEnv?: Readonly<Record<string, string>>
}

export async function execute(input: ExecuteInput): Promise<CommandResult> {
  const env = isolatedEnv(input.extraEnv)
  return new Promise<CommandResult>((resolve, reject) => {
    const child = spawn(input.command, input.args, {
      shell: false,
      env,
      stdio: ["ignore", "pipe", "pipe"]
    })
    const chunks = createOutputCollector(input.config.maxOutputBytes)
    let timedOut = false
    const timeout = setTimeout(() => {
      timedOut = true
      child.kill("SIGTERM")
    }, input.config.commandTimeoutMs)

    child.stdout.on("data", (chunk: Buffer) => chunks.stdout.push(chunk))
    child.stderr.on("data", (chunk: Buffer) => chunks.stderr.push(chunk))
    child.on("error", (error: NodeJS.ErrnoException) => {
      clearTimeout(timeout)
      reject(new CommandExecutionError(input.command, error.message))
    })
    child.on("close", (code) => {
      clearTimeout(timeout)
      resolve({
        command: input.command,
        args: input.args,
        exitCode: code ?? 1,
        stdout: redact(chunks.stdout.text()),
        stderr: redact(chunks.stderr.text()),
        timedOut
      })
    })
  })
}

export function redact(value: string): string {
  return SECRET_PATTERNS.reduce((current, pattern) => current.replace(pattern, "[REDACTED]"), value)
}

function isolatedEnv(extraEnv: Readonly<Record<string, string>> = {}): NodeJS.ProcessEnv {
  const allowedKeys = [
    "PATH",
    "HOME",
    "AWS_PROFILE",
    "AWS_CONFIG_FILE",
    "AWS_SHARED_CREDENTIALS_FILE",
    "AWS_REGION",
    "AWS_DEFAULT_REGION",
    "KUBECONFIG",
    "EKS_FAKE_CLI_LOG",
    "HELM_CACHE_HOME",
    "HELM_CONFIG_HOME",
    "HELM_DATA_HOME"
  ] as const
  const nextEnv: NodeJS.ProcessEnv = {}
  for (const key of allowedKeys) {
    const value = process.env[key]
    if (value !== undefined) {
      nextEnv[key] = value
    }
  }
  for (const [key, value] of Object.entries(extraEnv)) {
    nextEnv[key] = value
  }
  return nextEnv
}

function createOutputCollector(maxBytes: number): { readonly stdout: Collector; readonly stderr: Collector } {
  return { stdout: new Collector(maxBytes), stderr: new Collector(maxBytes) }
}

class Collector {
  readonly #chunks: Buffer[] = []
  #bytes = 0
  #truncated = false

  constructor(readonly maxBytes: number) {}

  push(chunk: Buffer): void {
    const remaining = this.maxBytes - this.#bytes
    if (remaining <= 0) {
      this.#truncated = true
      return
    }
    const accepted = chunk.subarray(0, remaining)
    this.#chunks.push(accepted)
    this.#bytes += accepted.byteLength
    this.#truncated = this.#truncated || accepted.byteLength < chunk.byteLength
  }

  text(): string {
    const suffix = this.#truncated ? "\n[output truncated]" : ""
    return `${Buffer.concat(this.#chunks).toString("utf8")}${suffix}`
  }
}

File: safe_eks_mcp/exec.py
from __future__ import annotations

import os
import re
import subprocess
from collections.abc import Mapping
from typing import Final

from .types import CommandResult, RuntimeConfig

SECRET_PATTERNS: Final = (
    re.compile(r"AKIA[0-9A-Z]{16}"),
    re.compile(r"ASIA[0-9A-Z]{16}"),
    re.compile(r"(?<=aws_secret_access_key\s=)[^\s]+", re.IGNORECASE),
    re.compile(r"(?<=aws_session_token\s=)[^\s]+", re.IGNORECASE),
    re.compile(r"(?<=Authorization:\sBearer\s)[A-Za-z0-9._~+/=-]+", re.IGNORECASE),
    re.compile(r"(?<=token: )[A-Za-z0-9._~+/=-]+", re.IGNORECASE),
)
ENV_ALLOWLIST: Final = (
    "PATH",
    "HOME",
    "AWS_PROFILE",
    "AWS_CONFIG_FILE",
    "AWS_SHARED_CREDENTIALS_FILE",
    "AWS_REGION",
    "AWS_DEFAULT_REGION",
    "KUBECONFIG",
    "EKS_FAKE_CLI_LOG",
    "HELM_CACHE_HOME",
    "HELM_CONFIG_HOME",
    "HELM_DATA_HOME",
)


def execute(command: str, args: tuple[str, ...], config: RuntimeConfig) -> CommandResult:
    timeout_seconds = config.command_timeout_ms / 1000
    try:
        completed = subprocess.run(
            (command, *args),
            check=False,
            capture_output=True,
            env=isolated_env(os.environ),
            shell=False,
            text=True,
            timeout=timeout_seconds,
        )
    except subprocess.TimeoutExpired as error:
        return CommandResult(
            command=command,
            args=args,
            exit_code=1,
            stdout=redact(truncate(timeout_output(error.stdout), config.max_output_bytes)),
            stderr=redact(truncate(timeout_output(error.stderr), config.max_output_bytes)),
            timed_out=True,
        )
    except OSError as error:
        return CommandResult(
            command=command,
            args=args,
            exit_code=1,
            stdout="",
            stderr=redact(str(error)),
            timed_out=False,
        )
    return CommandResult(
        command=command,
        args=args,
        exit_code=completed.returncode,
        stdout=redact(truncate(completed.stdout, config.max_output_bytes)),
        stderr=redact(truncate(completed.stderr, config.max_output_bytes)),
        timed_out=False,
    )


def isolated_env(source: Mapping[str, str]) -> dict[str, str]:
    return {key: value for key in ENV_ALLOWLIST if (value := source.get(key)) is not None}


def truncate(value: str, max_bytes: int) -> str:
    encoded = value.encode("utf-8")
    if len(encoded) <= max_bytes:
        return value
    return encoded[:max_bytes].decode("utf-8", errors="ignore") + "\n[output truncated]"


def redact(value: str) -> str:
    current = value
    for pattern in SECRET_PATTERNS:
        current = pattern.sub("[REDACTED]", current)
    return current


def timeout_output(value: str | bytes | None) -> str:
    if value is None:
        return ""
    if isinstance(value, bytes):
        return value.decode("utf-8", errors="ignore")
    return value
